docs/src/design/programs.md

```diff
@@ -0,0 +1,98 @@
+
+From the user's perspective, all operations are executed in order, however, the VM may insert occasional `NOOP`s to ensure proper alignment of all operations in the sequence. Currently, the alignment requirements are as follows:
+* An operation carrying an immediate value cannot be the first operation in a group. Thus, for example, if a `PUSH` operation is the last operation in a group, the VM will insert a `NOOP` after it.
+
+A *span* block does not have any children, and thus, must be leaf node in the tree.
+
+## Program example
+Consider the following program, where $a_0, ..., a_i$, $b_0, ..., b_j$ etc. represent individual operations:
+
+```
+a_0, ..., a_i
+if.true
+    b_0, ..., b_j
+else
+    c_0, ..., c_k
+    while.true
+        d_0, ..., d_n
+    end
+    e_0, ..., e_m
+end
+f_0, ..., f_l
+```
+
+A MAST for this program would look as follows:
+
+![](https://i.imgur.com/hEWuQnG.png)
+
+Execution of this program would proceed as follows:
+
+1. The VM will start execution at the root of the program which is block $B_5$.
+2. Since, $B_5$ is a *join block*, the VM will attempt to execute block $B_4$ first, and only after that execute block $f$.
+3. Block $B_4$ is also a *join block*, and thus, the VM will execute block $a$ by executing operations $a_0, ..., a_i$ in sequence, and then execute block $B_3$.
+4. Block $B_3$ is a *split block*, and thus, the VM will pop the value off the top of the stack. If the popped value is $1$, operations from block $b$ will be executed in sequence. If the popped value is $0$, then the VM will attempt to execute block $B_2$.
+5. $B_2$ is a *join block*, thus, the VM will try to execute block $B_1$ first, and then execute operations from block $e$.
+6. Block $B_1$ is also a *join_block*, and thus, the VM will first execute all operations in block $c$, and then will attempt to execute block $B_0$.
+7. Block $B_0$ is a loop block, thus, the VM will pop the value off the top of the stack. If the pooped value is $1$, the VM will execute the body of the loop defined by block $d$. If the popped value is $0$, the VM will not execute block $d$ and instead will move up the tree executing first block $e$, then $f$.
+8. If the VM does enter the loop, then after operation $d_n$ is executed, the VM will pop the value off the top of the stack again. If the popped value is $1$, the VM will execute block $d$ again, and again until the top of the stack becomes $0$. Once the top of the stack becomes $0$, the VM will exit the loop and will move up the tree executing first block $e$, then $f$.
+
+## Program hash computation
+Every Miden VM program can be reduced to a unique hash value. Specifically, it is infeasible to find two Miden VM programs with distinct semantics which hash to the same value. Padding a program with `NOOP`s does not change a program's execution semantics, and thus, programs which differ only in the number and/or placement of `NOOP`s may hash to the same value.
```

Comment:
Padding with `NOOP`s can affect the hash value sometimes though, right? I think it would be good to note that explicitly (e.g., "...may hash to the same value, although this is not guaranteed.")

Reply:
Updated! (with a slightly different wording).